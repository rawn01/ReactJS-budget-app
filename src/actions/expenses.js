import uuid from 'uuid';

// Add Expense action generator 
export const addExpense = ({desc = "", note = "", amount = 0, createdAt = 0} = {}) => {
   return {
      type: 'ADD_EXPENSE',
      expense: {
         id: uuid(),
         description: desc,
         note: note,
         amount: amount,
         createdAt: createdAt
      }
   }
};

// Remove Expense action generator
export const removeExpense = ({ id }) => {
   return {
      type: 'REMOVE_EXPENSE',
      id: id
   };
};

// Edit Expense action generator
export const editExpense = (id, updates) => {
   return {
      type: 'EDIT_EXPENSE',
      id: id,
      updates: updates
   }
};
