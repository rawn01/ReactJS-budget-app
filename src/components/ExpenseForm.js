import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


class ExpenseForm extends React.Component {
   constructor(props) {
      super(props);
      this.state = {
         description: props.expense ? props.expense.description : '',
         note: props.expense ? props.expense.note : '',
         amount: props.expense ? props.expense.amount.toString() : '',
         createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
         focused: false,
         error: ''
      };
      // console.log("waka-waka", this.state);
   }

   onDescriptionChange = (e) => {
      const description = e.target.value;
      this.setState(() => {
         return {
            description: description
         };
      });
   };

   onNoteChange = (e) => {
      const note = e.target.value;
      this.setState(() => {
         return {
            note: note
         };
      });
   };

   onAmountChange = (e) => {
      const amount = e.target.value;
      if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
         this.setState(() => {
            return {
               amount: amount
            }
         });
      }
   };

   onDateChange = (createdAt) => {
      if(createdAt) {
         this.setState(() => {
            return {
               createdAt: createdAt
            }
         });
      }
   };

   onFocusChange = ({ focused }) => {
      this.setState(() => {
         return {
            focused: focused
         };
      })
   };

   onFormSubmit = (e) => {
      e.preventDefault();

      if(!this.state.description || !this.state.amount) {
         // Set error state
         this.setState(() => {
            return {
               error: "Please provide description and amount"
            };
         })
      } else {
         // Clear the error
         this.setState(() => {
            return {
               error: ''
            }
         });
         
         this.props.onSubmit({
            desc: this.state.description,
            amount: parseFloat(this.state.amount),
            createdAt: this.state.createdAt.valueOf(),
            note: this.state.note
         });
      }
   }

   render() {
      return (
         <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onFormSubmit}>
               <input type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.onDescriptionChange} />
               <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange} />
               <SingleDatePicker date={this.state.createdAt} onDateChange={this.onDateChange} focused={this.state.focused} onFocusChange={this.onFocusChange} numberOfMonths={1} isOutsideRange={() => false} />
               <textarea placeholder="Add a note (optional)" value={this.state.note} onChange={this.onNoteChange}></textarea>
               <button>Add Expense</button>
            </form>
         </div>
      );
   }
};

export default ExpenseForm;