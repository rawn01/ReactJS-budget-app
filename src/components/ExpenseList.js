import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
   return (
      <div>
         <h1>Expense List</h1>
         {
            props.expenses.length === 0 ? (
               <p>No expenses</p>
            ) : (
               props.expenses.map((item) => {
                  return <ExpenseListItem key={item.id} {...item} />
               })
            )
         }
      </div>
   );
};

const mapStateToProps = (state) => {
   return {
      expenses: selectExpenses(state.expenses, state.filters)
   };
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);


export default ConnectedExpenseList;



// const ConnectedExpenseList = connect((state) => {
//    return {
//       expenses: state.expenses
//    };
// })(ExpenseList);


// export default ConnectedExpenseList;