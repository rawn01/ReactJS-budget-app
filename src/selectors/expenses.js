import moment from 'moment';

// Get Visible Expenses

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
   return expenses.filter((item) => {
      // const createdAtMoment = moment(item.createdAt);
      // console.log(createdAtMoment);
      // console.log(item.createdAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(item.createdAt, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(item.createdAt, 'day') : true;
      const textMatch =  item.description.toLowerCase().includes(text.toLowerCase());
      
      return startDateMatch && endDateMatch && textMatch;
   }).sort((a, b) => {
      if(sortBy === "date") {
         return a.createdAt < b.createdAt ? 1 : -1;
      }
      if(sortBy === "amount") {
         return a.amount > b.amount ? -1 : 1;
      }
   });
};

export default getVisibleExpenses;