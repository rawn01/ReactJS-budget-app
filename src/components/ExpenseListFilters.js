import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {
   constructor(props) {
      super(props);
      this.state = {
         focused: null
      };
   }

   onDatesChange = ({ startDate, endDate }) => {
      this.props.setStartDate(startDate);
      this.props.setEndDate(endDate);
   };

   onFocusChange = (focused) => {
      this.setState(() => {
         return {
            focused: focused
         };
      })
   };

   onTextChange = (e) => {   
      this.props.setTextFilter(e.target.value);
   };

   onSortChange = (e) => {
      if(e.target.value === 'date') {
         this.props.sortByDate();
      } else if(e.target.value === "amount") {
         this.props.sortByAmount();
      }
   };

   render() {
      return (
         <div>
            <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />

            <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
               <option value="date" >Date</option>
               <option value="amount">Amount</option>
            </select>

            <DateRangePicker 
               startDate={this.props.filters.startDate}
               endDate={this.props.filters.endDate}
               onDatesChange={this.onDatesChange}
               focusedInput={this.state.focused}
               onFocusChange={this.onFocusChange}
               startDateId={'start'}
               endDateId={'end'}
               numberOfMonths={1}
               isOutsideRange={() => false}
               showClearDates={true}
            />
         </div>
      );
   }
}

const mapStateToProps = (state) => {
   return {
      filters: state.filters
   };
};

const matchDispatchToProps = (dispatch) => {
   return {
      setTextFilter: (text) => {
         return dispatch(setTextFilter(text));
      },
      sortByDate: () => {
         return dispatch(sortByDate());
      },
      sortByAmount: () => {
         return dispatch(sortByAmount());
      },
      setStartDate: (startDate) => {
         return dispatch(setStartDate(startDate));
      },
      setEndDate: (endDate) => {
         return dispatch(setEndDate(endDate));
      }
   };
};

const ConnectedExpenseListFilters = connect(mapStateToProps, matchDispatchToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;