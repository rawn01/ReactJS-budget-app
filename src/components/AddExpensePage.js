import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export const AddExpensePage = (props) => {
   return (
      <div>
         <h1>Add Expense</h1>
         <ExpenseForm onSubmit={(expense) => {
            // props.dispatch(addExpense(expense));
            props.addExpense(expense);
            props.history.push('/');
         }} />
      </div>
   );
};

// Used for unit-TESTING
const mapDispatchToProps = (dispatch) => {
   return {
      addExpense: (expense) => {
         return dispatch(addExpense(expense));
      }
   }
};

const ConnectedAddExpensePage = connect(undefined, mapDispatchToProps)(AddExpensePage);

export default ConnectedAddExpensePage;