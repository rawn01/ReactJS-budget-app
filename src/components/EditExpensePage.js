import React from 'react';
import { connect } from 'react-redux';

import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export const EditPage = (props) => {
   return (
      <div>
         <ExpenseForm expense={props.expense} onSubmit={(expense) => {
            props.editExpense(props.expense.id, expense);
            props.history.push('/');
         }} />
         <button onClick={() => {
            props.removeExpense({ id: props.expense.id });
            props.history.push('/');
         }}>Remove</button>
      </div>
   );   
};

const mapStateToProps = (state, props) => {
   return {
      expense: state.expenses.find((expense) => {
         return props.match.params.id === expense.id;
      })
   }
};

// Used for TESTING
const mapDispatchToProps = (dispatch) => {
   return {
      editExpense: (id, expense) => {
         return dispatch(editExpense(id, expense));
      },
      removeExpense: (data) => {
         return dispatch(removeExpense(data));
      }
   };
};

const ConnectedEditPage = connect(mapStateToProps, mapDispatchToProps)(EditPage);

export default ConnectedEditPage;