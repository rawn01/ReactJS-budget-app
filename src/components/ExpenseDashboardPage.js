import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => {
   return (
      <div>
         <p>This is from ExpenseDashboardPage!</p>
         <ExpenseListFilters />
         <ExpenseList />
      </div>
   );
};

export default ExpenseDashboardPage;