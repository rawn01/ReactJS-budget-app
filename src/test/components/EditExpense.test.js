import React from 'react';
import { shallow } from 'enzyme';

import expenses from '../fixtures/expenses';
import { EditPage } from '../../components/EditExpensePage';

test('should render the edit expense page', () => {
   const editExpense = jest.fn();
   const removeExpense = jest.fn();
   const history = {
      push: jest.fn()
   };
   const wrapper = shallow(
      <EditPage 
         editExpense={editExpense}
         removeExpense={removeExpense}
         history={history}
         expense={expenses[0]}
      />
   );

   expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
   const editExpense = jest.fn();
   const removeExpense = jest.fn();
   const history = {
      push: jest.fn()
   };
   const wrapper = shallow(
      <EditPage 
         editExpense={editExpense}
         removeExpense={removeExpense}
         history={history}
         expense={expenses[2]}
      />
   );

   wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
   expect(history.push).toHaveBeenLastCalledWith('/');
   expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});

test('should handle remove expense', () => {
   const editExpense = jest.fn();
   const removeExpense = jest.fn();
   const history = {
      push: jest.fn()
   };
   const wrapper = shallow(
      <EditPage 
         editExpense={editExpense}
         removeExpense={removeExpense}
         history={history}
         expense={expenses[0]}
      />
   );

   wrapper.find('button').prop('onClick')(expenses[0]);
   expect(history.push).toHaveBeenLastCalledWith('/');
   expect(removeExpense).toHaveBeenLastCalledWith({
      id: expenses[0].id
   });
});