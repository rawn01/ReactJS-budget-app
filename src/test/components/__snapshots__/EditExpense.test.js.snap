// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the edit expense page 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 1,
        "createdAt": 0,
        "description": "Toffee",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
